conseil-lorre/src/main/scala/tech/cryptonomic/conseil/indexer/tezos/ConsistentForkDetector.scala

```diff
@@ -0,0 +1,109 @@
+
+/* Common types and utils */
+object ConsistentForkDetector {
+
+  case class ConsistentDetectionRepeatingFailure(message: String, attempts: Int) extends RuntimeException(message)
+
+}
+
+/** Provides guarantees of a more consistent fork level, given
+  * a tezos node that might return unstable results - from different
+  * alternating forks - during the search.
+  * The returned value is guaranteed to be locally consistent, i.e.
+  * local blocks are consistent before the fork level and with the
+  * new remote block from the fork on.
+  *
+  * @param indexerSearch same as for the generic [[ForkDetector]]
+  * @param nodeSearch same as for the generic [[ForkDetector]]
+  * @param nodeData will provide [[BlockData]] from the node for a given level
+  * @param maxAttempts how many times to try upon inconsistent fork level detection, has to be strictly positive
+  */
+class ConsistentForkDetector[Eff[_]: Monad](
+    indexerSearch: SearchBlockId[Eff, TezosBlockHash],
+    nodeSearch: SearchBlockId[Eff, TezosBlockHash],
+    nodeData: SearchBlockData[Eff],
+    maxAttempts: Int
+) extends ForkDetector[Eff, TezosBlockHash](indexerSearch, nodeSearch)
+    with LazyLogging {
+
+  /* validate early */
+  assert(
+    maxAttempts > 0,
+    s"Please provide the ForkDetector with a positive value for number of attempts, when results are inconsistent. The max-attempts given was $maxAttempts"
+  )
+
+  /** See [[ForkDetector#searchForkLevel]] */
+  override def searchForkLevel(low: Long, high: Long): Eff[Long] = {
+    /* we capture the original search and add post consistency check
+     * with recursive retries in case of failure
+     */
+    val runRegularSearch = () => super.searchForkLevel(low, high)
+
+    def attemptSearch(attemptsLeft: Int): Eff[Long] = {
+      if (attemptsLeft == 0)
+        throw new ConsistentDetectionRepeatingFailure(
+          "I've not been able to detect the block level at which the fork happened. Any attempt gave inconsistent results between local and remote data.",
+          maxAttempts
+        )
+      logger.warn(
+        s"Looking for the block where the local indexer diverged from the chain node due to a fork. I will make $attemptsLeft attempt(s)."
+      )
+      val candidate = runRegularSearch()
+      val consistencyCheck = candidate.flatMap(detectionIsLocallyConsistent)
+      consistencyCheck.ifM(ifTrue = candidate, ifFalse = attemptSearch(attemptsLeft - 1))
+    }
+
+    attemptSearch(maxAttempts)
+  }
+
+  /** We want to make sure that during the search execution, which probably handles
+    * multiple I/O calls to compare indexer and node results, the node has kept giving
+    * consistent results.
+    * It might happen that network calls to both services were not fast enough, to the
+    * point that the fork on which the reference node was at the time of detection changed
+    * during the algorthm execution. This might be a reasonable expectation in a highly
```

Comment:
Typo: Algorithm

Reply:
Fixed, and I additionally rephrased the "locally consistent" definition, trying to make it easier to understand.